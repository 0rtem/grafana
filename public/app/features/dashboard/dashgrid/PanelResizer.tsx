import React, { PureComponent } from 'react';
import { ResizableBox } from 'react-resizable';
import { PanelModel } from '../panel_model';
import { debounce } from 'lodash';

interface DimensionsXY {
  x: number;
  y: number;
}

export interface Props {
  isEditing: boolean;
  panel: PanelModel;
}

export interface State {
  panelDimensions: DimensionsXY;
  hasEventListener: boolean;
}

export class PanelResizer extends PureComponent<Props, State> {
  element: HTMLElement;
  debouncedResizeDone: () => void;
  debouncedUpdatePanelDimensions: () => void;

  constructor(props) {
    super(props);
    const { panel } = this.props;
    const debounceTimeMs = 100;
    this.debouncedResizeDone = debounce(() => {
      panel.resizeDone();
    }, debounceTimeMs);

    this.debouncedUpdatePanelDimensions = debounce(this.updatePanelDimensions, debounceTimeMs);

    this.state = {
      panelDimensions: this.panelDimensions,
      hasEventListener: false,
    };
  }

  componentDidMount() {
    this.handleResizeEventListener();
  }

  componentDidUpdate() {
    this.handleResizeEventListener();
  }

  componentWillUnmount() {
    this.removeEventListener();
  }

  handleResizeEventListener = () => {
    const { isEditing } = this.props;
    const { hasEventListener } = this.state;
    if (isEditing && !hasEventListener) {
      this.addEventListener();
    } else if (!isEditing && hasEventListener) {
      this.removeEventListener(true);
    }
  };

  addEventListener = () => {
    window.addEventListener('resize', this.debouncedUpdatePanelDimensions);
    this.setState({ hasEventListener: true });
  };

  removeEventListener = (skipStateChange = false) => {
    window.removeEventListener('resize', this.debouncedUpdatePanelDimensions);
    if (!skipStateChange) {
      this.setState({ hasEventListener: false });
    }
  };

  get panelDimensions(): DimensionsXY {
    return {
      x: document.documentElement.scrollWidth,
      y: Math.floor(document.documentElement.scrollHeight * 0.4),
    };
  }

  updatePanelDimensions = () => {
    this.setState({
      panelDimensions: this.panelDimensions,
    });
  };

  render() {
    const { children, isEditing } = this.props;
    const { x, y } = this.state.panelDimensions;
    return isEditing ? (
      <ResizableBox height={y} width={x} axis="y" onResize={this.debouncedResizeDone}>
        <>{children}</>
      </ResizableBox>
    ) : (
      <>{children}</>
    );
  }
}
