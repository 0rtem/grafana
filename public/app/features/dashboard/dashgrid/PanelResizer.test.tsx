import React from 'react';
import renderer from 'react-test-renderer';
import { PanelResizer } from './PanelResizer';
import { PanelModel } from '../panel_model';

const panel = new PanelModel({ collapsed: false });

describe('PanelResizer', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <PanelResizer isEditing={true} panel={panel}>
          <p>PanelResizer</p>
        </PanelResizer>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
