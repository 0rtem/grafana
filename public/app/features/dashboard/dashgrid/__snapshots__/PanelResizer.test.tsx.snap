// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelResizer renders correctly 1`] = `
<div
  className="react-resizable"
  style={
    Object {
      "height": "0px",
      "width": "0px",
    }
  }
>
  <p>
    PanelResizer
  </p>
  <span
    className="react-resizable-handle"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "touchAction": "none",
      }
    }
  />
</div>
`;
